[WIP] Add collections endpoint to API v2

barbour-em opened the pull request:
## Purpose 
Create endpoint for collections due to restricted feature set
This closes #3047, #3167 and complies with the changes made in #3118 


## Changes
Create api/collections directory similar to that of api/nodes but with stricter serialization. 
Create, update, and delete for both folders as well as their pointers
Cannot update your Dashboard
Smart folders for all my projects and all my registrations are viewable in `collections/dashboard/children` as well as in `/collections/~amp/` and `/collections/~amr/` respectively 
Update url routes to all have trailing slashes
View tests for this end point are based on `API_BASE` 

The json returned from this endpoint looks like
```json
{
    "data": [
        {
            "id": "23znr",
            "title": "Chopsticks",
            "date_created": "2015-06-12T16:12:04.420000",
            "date_modified": "2015-06-12T16:12:50.752000",
            "links": {
                "pointers": {
                    "count": 0,
                    "related": "https://localhost:8000/v2/collections/23znr/pointers/"
                },
                "self": "https://localhost:8000/v2/collections/23znr/",
                "children": {
                    "count": 0,
                    "related": "https://localhost:8000/v2/nodes/23znr/children/"
                }
            },
            "properties": {
                "dashboard": false,
                "collection": false,
                "registration": false
            },
        },
    ]
}
```
This differs from the typical node request in that collections do not return any information about files, contributors, registrations, or the html link 

Changes to the documentation / Swagger UI as well as the routes created:
![screen shot 2015-06-19 at 9 32 13 am](https://cloud.githubusercontent.com/assets/8905795/8254423/59228cee-1666-11e5-9cdb-c7594e18d709.png)

Swagger UI documentation has been updated to reflect the differences in the collection endpoint:
![screen shot 2015-06-19 at 9 40 22 am](https://cloud.githubusercontent.com/assets/8905795/8254530/42cff98a-1667-11e5-8f37-625f0e4bd51a.png)

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,716 @@
+        res = self.app.get(self.url)
+        node_json = res.json['data']
+
+        ids = [each['id'] for each in node_json]
```
Minor, but I prefer to use descriptive nouns instead of each. Since it's a comprehension, each is implied. Something like `[node['id'] for node in node_json]` will give better context for people reading through.

barbour-em replied:
Done ✔

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,716 @@
+        self.basic_non_contrib_auth = (self.non_contrib.username, 'justapoorboy')
+
+        self.deleted = ProjectFactory(is_deleted=True, is_folder=True)
+        self.private = ProjectFactory(is_public=False, creator=self.user, is_folder=True)
```
Go ahead and use the FolderFactory rather than setting is_folder to true.

barbour-em replied:
Not sure why I didn't do that in the first place. Thanks for the reminder

brianjgeiger commented:
Okay, there was one more kind of big thing that I've just noticed, sorry. As things stand, there is no such thing as a public collection, and collections can currently only be owned by a single person. The second is more likely to change in the near future than the first.

In general terms, what this means is that the default query should look for nodes that are owned by the current user, and that the permission classes should not expect public folders or any sort of read-only access. If you want, I can go through and point out some things that will change based on this, or you can do a pass first and let me look through after.

I will try to see if there are any more differences between normal nodes and collections during this pass. Hopefully no.

brianjgeiger commented:
Be sure to start by modifying your tests first, then change the code.

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,716 @@
+        ApiTestCase.tearDown(self)
+        Node.remove()
+
+    def test_get_all_projects_with_no_filter_logged_in(self):
```
Change the names for all the tests from using `projects` to using `collections` (unless it is actually referring to a project).

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,716 @@
+        self.folder = FolderFactory()
+        self.dashboard = DashboardFactory()
+
+        self.url = "/v2/collections/"
```
This will be happening independently for existing tests, but see #3062 to make this more futureproof.

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,716 @@
+        self.user_two.set_password('justapoorboy')
+        self.user_two.save()
+        self.basic_auth_two = (self.user_two.username, 'justapoorboy')
+        self.project_one = ProjectFactory(title="Project One", is_public=True, is_folder=True)
```
Also variables, if it's a collection, call it a collection

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,716 @@
+        assert_not_in(self.folder._id, ids)
+        assert_not_in(self.dashboard._id, ids)
+
+    def test_get_all_projects_with_no_filter_not_logged_in(self):
```
Should be a list of 0 items

brianjgeiger commented:
Looks like we'll need tests for smart folders, as well. Ensure that they are getting what we expect.

barbour-em commented:
Addressed the above issues. We will need to discuss how we want to test smart folders based on where we want to display their information (i.e. in Dashboard Detail vs say a Collection List or Detail). The Travis failure looks like a problem with invoke and requirements rather than my code but I could be so I will leave that to be reviewed. Other than that, ready for another pass!

brianjgeiger commented on api/collections/serializers.py:
```diff
@@ -0,0 +1,113 @@
+            'collection': obj.is_folder,
+            'dashboard': obj.is_dashboard,
+            'smart_folder': obj.smart_folder,
+            'modified_by': modified_by,
```
The help text on Properties indicates that these are all booleans. So we either need to move modified_by out of properties, or we need to change the help text everywhere. I think go ahead and move it out to its own field, as we may have to do other large-scale restructuring of all the data, and no sense going to crazy right now.

brianjgeiger commented on api/collections/serializers.py:
```diff
@@ -0,0 +1,113 @@
+    modified_by = ser.CharField(read_only=True, source='get_modified_by')
+    links = CollectionLinksField({
+        'children': {
+            'related': Link('collections:collection-children', kwargs={'pk': '<pk>'}),
```
This is happening elsewhere for existing routes, but go ahead and change `pk` to `collection_id`. This will happen in your urls file as well (and I'll likely comment on that when I get there). This will make documentation better.

brianjgeiger commented on api/collections/urls.py:
```diff
@@ -0,0 +1,15 @@
+    # url(r'^blog/', include('blog.urls')),
+    url(r'^$', views.CollectionList.as_view(), name='collection-list'),
+    url(r'^dashboard/$', views.DashboardDetail.as_view(), name='dashboard-detail'),
+    url(r'^(?P<pk>\w+)/$', views.CollectionDetail.as_view(), name='collection-detail'),
```
Change `pk` throughout this to `collection_id`.

brianjgeiger commented:
Could you pull from develop? That should fix the build issue with Travis.

brianjgeiger commented on api/collections/views.py:
```diff
@@ -0,0 +1,246 @@
+    serializer_class = CollectionPointersSerializer
+
+    def get_queryset(self):
+        pointers = self.get_node().nodes_pointer
```
It looks like we aren't actually performing any of the queries for the smart folders. We'll have to check to see if this is a smart folder, and then we'll need to do the smart folder query here if it is. That will likely mean moving the list of smart folders to somewhere that all the classes can use and turning it into a dictionary so that we can store the query along with the smart folders themselves.

brianjgeiger commented on api/collections/views.py:
```diff
@@ -0,0 +1,246 @@
+
+    # overrides ListAPIView
+    def get_queryset(self):
+        nodes = self.get_node().nodes
```
This is likely to fail on smart folders, so you might have to check to see if this is a smart folder here and return an empty list.

brianjgeiger commented:
Pass done. :penguin: Some small things, implementation of Smart Collections, and tests.

brianjgeiger commented:
The unused FolderFactory import is breaking travis.

barbour-em commented:
I saw, shame on me! Will be removed in the next commit

barbour-em commented:
@brianjgeiger this is ready for another pass

brianjgeiger commented on api/collections/serializers.py:
```diff
@@ -0,0 +1,126 @@
+
+    def get_node_count(self, obj):
+        if isinstance(obj, dict) and obj['properties']['smart_folder'] is True:
+            return ''
```
Smart folders don't have children, so it's safe to return 0 here. I haven't run this yet, but we'd want the count to show as 0 rather than empty.

brianjgeiger commented on api/collections/serializers.py:
```diff
@@ -0,0 +1,126 @@
+
+    def get_pointers_count(self, obj):
+        if isinstance(obj, dict) and obj['properties']['smart_folder'] is True:
+            return ''
```
This one would be better if we could get the count of items in the smart folder. I don't know if we can do it in a resource-efficient way, though.

brianjgeiger commented on api/collections/serializers.py:
```diff
@@ -0,0 +1,126 @@
+
+    def get_modified_by(self, obj):
+        if isinstance(obj, dict) and obj['properties']['smart_folder'] is True:
+            return 'Admin'
```
I think the current user name would be appropriate here or leave it blank (which is what the current version does)

brianjgeiger commented on api/collections/serializers.py:
```diff
@@ -0,0 +1,126 @@
+        modified_by = user.family_name or user.given_name
+        return modified_by
+
+    def get_user_auth(self, request):
```
I think the import problem has been fixed, so this could totally be moved to base/utils now.

brianjgeiger commented on api/collections/serializers.py:
```diff
@@ -0,0 +1,126 @@
+    def create(self, validated_data):
+        node = Node(**validated_data)
+        node.save()
+        node.is_folder = True
```
Would this be better above the node.save()?

brianjgeiger commented on api/collections/urls.py:
```diff
@@ -0,0 +1,15 @@
+    # url(r'^blog/', include('blog.urls')),
+    url(r'^$', views.CollectionList.as_view(), name='collection-list'),
+    url(r'^dashboard/$', views.DashboardDetail.as_view(), name='dashboard-detail'),
+    url(r'^[~]?(?P<pk>\w+)/$', views.CollectionDetail.as_view(), name='collection-detail'),
```
We shouldn't be using <pk> anywhere anymore. We should be using <collection_id> instead. See #3121 for the most specific changes you'll need to make to urls, serializers, and views.

brianjgeiger commented on api/collections/views.py:
```diff
@@ -0,0 +1,283 @@
+from website.views import find_dashboard
+
+smart_folders = {
+    '~amp': Q('category', 'eq', 'project') & Q('is_deleted', 'eq', False) &
```
Minor formatting issue, but let's put each Q on a different line. I think that will make it easier to follow with this many criteria.

brianjgeiger commented on api/collections/views.py:
```diff
@@ -0,0 +1,283 @@
+            ret = {
+                'id': '{}'.format(key),
+                'pk': key,
+                'title': "Smart Folder {}".format(key),
```
A smart folder is still a folder, so include the is_folder property here as well. I presume if is_dashboard exists, it will return falsy in the serializer, so if it doesn't have a problem, that's probably fine.

brianjgeiger commented on api/collections/urls.py:
```diff
@@ -0,0 +1,15 @@
+    # url(r'^$', 'api.views.home', name='home'),
+    # url(r'^blog/', include('blog.urls')),
+    url(r'^$', views.CollectionList.as_view(), name='collection-list'),
+    url(r'^dashboard/$', views.DashboardDetail.as_view(), name='dashboard-detail'),
```
Thiiiis should not be necessary. You should be able to do this the same way you're doing the smart folders.

barbour-em replied:
You're right, should have been removed a while ago

brianjgeiger replied:
If it's a lot more complicated than the extra view, then go ahead and keep it, but you might be able to dump a bunch of lines of code now.

brianjgeiger commented on api/collections/views.py:
```diff
@@ -0,0 +1,283 @@
+        serializer.save(creator=user)
+
+
+class DashboardDetail(generics.ListCreateAPIView, ODMFilterMixin):
```
Similarly, this should be able to go away.

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,622 @@
+        assert_in(self.collection._id, ids)
+
+    def test_return_collection_list_logged_out_user(self):
+        res = self.app.get(self.url, auth=self.basic_auth)
```
This is not testing a logged out user. It passes because the assert is incorrect.

barbour-em replied:
This test should check for a 403 instead correct?

brianjgeiger replied:
Yeah, that sounds right.

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,622 @@
+        ids = [node['id'] for node in res.json['data']]
+        assert_in(self.collection._id, ids)
+
+        Node.remove()
```
There should be a tearDown method, and this should be in that method.

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,622 @@
+        self.user_two.set_password('justapoorboy')
+        self.user_two.save()
+        self.basic_auth_two = (self.user_two.username, 'justapoorboy')
+        self.collection_one = FolderFactory(title="Project One", is_public=True, creator=self.user_one)
```
Public collections aren't really a thing. Let's remove the concept of a public collection from these tests.

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,622 @@
+        self.collection_one = FolderFactory(title="Project One", is_public=True, creator=self.user_one)
+        self.collection_two = FolderFactory(title="Project Two", is_public=True, creator=self.user_one)
+        self.collection_three = FolderFactory(title="Three", is_public=True, creator=self.user_one)
+        self.private_collection_user_one = FolderFactory(title="Private Project User One", is_public=False,
```
Collections should be private by default, so you shouldn't need the is_public in here.

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,622 @@
+        node_json = res.json['data']
+
+        ids = [node['id'] for node in node_json]
+        assert_in(self.collection_one._id, ids)
```
If you are not logged in, the collections endpoint should return an empty list.

brianjgeiger replied:
No, sorry, I misread. Carry on. This is fine.

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,622 @@
+        assert_not_in(self.private_collection_user_two._id, ids)
+        assert_not_in(self.folder._id, ids)
+        assert_not_in(self.dashboard._id, ids)
+
```
Add a test for not being logged in to ensure that it returns an empty list.

barbour-em replied:
We discussed that a user must be logged in to get the list and the documentation was changed to reflect this. Is this no longer the case?

brianjgeiger replied:
No, that's probably good. I just forgot.

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,622 @@
+        self.user_two.save()
+        self.basic_auth_two = (self.user_two.username, 'justapoorboy')
+
+        self.public_folder = {'title': self.title, 'public': True, 'is_folder': True}
```
Again, currently no such thing as a public collection, so remove these from the tests, and don't override the is_folder property in these tests. (It *should* be protected in the model, but it's not, so we have to just make sure the view doesn't allow it)

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,622 @@
+        self.user_two.save()
+        self.basic_auth_two = (self.user_two.username, 'justapoorboy')
+
+        self.public_collection = FolderFactory(title=self.title, is_public=False, creator=self.user)
```
Blah blah public private blah collections blah.

brianjgeiger commented on tests/api_tests/collections/test_views.py:
```diff
@@ -0,0 +1,622 @@
+        # presenting a basic authentication dialog box in the front end. We may change this as we understand CAS
+        # a little better
+        assert_equal(res.status_code, 403)
+
```
And a test for smart folders returning the proper information in their pointers list as well.

brianjgeiger commented:
Pass done. :penguin: Some little things, a general cleaning of tests, and smart folder detail tests.

brianjgeiger commented:
Oh, and it looks like your "be cool with a ~ in the pk" regex fix has some side effects, so you might need to fix it in a different way. 

![screen shot 2015-06-18 at 2 20 27 pm](https://cloud.githubusercontent.com/assets/6599111/8238711/6cca378e-15c5-11e5-9056-7da03c5bfb0e.png)

brianjgeiger commented:
It occurs to me that the node id for the smart folders in v2 doesn't have to match the node id for smart folders in v1. Go ahead to change those to be just amp and amr and take out your regex.

sloria commented:
@brianjgeiger Status update on this?

brianjgeiger commented:
@barbour-em got pulled off onto the pre-reg priority. After that's done, hopefully he can finish up on this.

barbour-em commented:
@sloria @brianjgeiger Correct, IIRC test coverage is the main thing left on this

barbour-em commented:
@brianjgeiger trying to get this knocked out, 2 tests for pointer collections are failing thought. It seems like `validated_data` isn't getting passed to the serializer `#create`. Any idea what could be going on here?

Also, I assume I will need to make some changes to the response format; ping @pattisdr on this. From what I understand this will include a new `relationships` LinkField containing children. Are pointers a relationship or do they stay in `links`?

brianjgeiger commented:
Pointers have had their name changed, so that could be what's causing that trouble. Don't worry about the changes based on @pattisdr's stuff yet, but it will be hitting in the next week-ish I would imagine. Hopefully it won't be too many changes to handle that.

pattisdr commented:
In my opinion, pointers are a relationship.

brianjgeiger commented:
Oh, and yes, pointers are a relationship (as Dawn said).

sloria commented:
closing in favor of https://github.com/CenterForOpenScience/osf.io/pull/4503
